### DISCUSSION TITLE
Axaml Previewer (LINUX - Proof of Concept)

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#7058] [General]
Hi,

I'd thought I'd share this as I've just got it working. There is a very good AXAML previewer extension for Visual Studio, however I'm primarily working with LINUX and thought there should be a previewer here also -- so decided to have a go at writing one.

What I'm planning is standalone previewer utility which should integrate with the "workflow" of any IDE. However, it is going to be standalone and not a plugin for Visual Code or other IDE. For this reason I am not using the HostApp/Previewer process, rather everything is being done with `AvaloniaRuntimeXamlLoader.Load()`.

Here is a screenshot of what I regard as proof of concept / R&D at the moment. It shows the preview of a MVVM window with bindings and a UserControl. To achieve this, the previewer has loaded the target assembly.

![image](https://user-images.githubusercontent.com/17156238/144255064-45df40db-bb7d-48fd-a799-f1f9a7b2f913.png)

The previewer will load the assembly where it is available, which will typically be the case, but it isn't always required to produce output. The screenshot below is a pure "code behind" window preview, but with no assembly loaded. Image assets are being located using the csproj file, not the assembly.

![image](https://user-images.githubusercontent.com/17156238/144251808-0d04ae3e-2b67-43b8-8e43-2f70cd7ffad0.png)

Here are a couple of test samples from the Avalonia sample ControlsCatalogue:

![image](https://user-images.githubusercontent.com/17156238/144286476-2b2d883e-46a6-4df0-a0fa-1476043241d3.png)

![image](https://user-images.githubusercontent.com/17156238/144288146-9a156b04-0977-4c38-8f28-b8309878b5d3.png)

**Comment by kekekeks** [comment]
You could have used the built-in web previewer (which is also used internally by Rider plugin).
```
dotnet exec --runtimeconfig ./path/to/app.runtimeconfig.json --depsfile  ./path/to/app.deps.json ~/.nuget/packages/avalonia/<avalonia-version>/tools/netcoreapp2.0/designer/Avalonia.Designer.HostApp.dll --transport file:///path/to/MainWindow.xaml --method html --html-url http://127.0.0.1:6001 ./path/to/app.dll
```

This will run previewer in auto-file-reload mode on http://127.0.0.1:6001

![gif](https://user-images.githubusercontent.com/1067584/69084594-a5e57e00-0a55-11ea-8152-a59834952166.gif)

**Reply by kekekeks** [reply]
There is also vscode plugin prototype: https://github.com/rstm-sf/view-avalonia-preview-vscode

**Comment by kekekeks** [comment]
>For this reason I am not using the HostApp/Previewer process, rather everything is being done with AvaloniaRuntimeXamlLoader.Load().

Note that previewer needs to apply styles from App.xaml, respect Avalonia version used by the target app and properly resolve libraries from nuget packages. In-process XAML loader can't really achieve that unless you do ALC shenanigans, manually parse .deps.json file and manually resolve nuget package paths.

**Comment by kuiperzone** [comment]
>respect Avalonia version used by the target app and properly resolve libraries from nuget packages.

I have done this actually.

**Comment by kuiperzone** [comment]
Hi @kekekeks , I was aware the PreviewerProcess as you kindly pointed it out to me previously. However, I had a different vision for my project which I want to follow. Thank you for your answers to my questions. It would be nice to think that you would answer them in the future, but I'm not sure whether I should count on that. In any case, I'm grateful. Andy

**Reply by kekekeks** [reply]
Yeah, sure, different applications need different approaches. We've chosen the current architecture because it was the most portable approach IDE-wise. We need to support Visual Studio, Rider and potentially VSCode. None of those runs .NET Core in their main IDE process. There are also considerations about isolation: previewed app shouldn't crash IDE process. But I can see scenarios when in-process preview would be desirable.

**Reply by kuiperzone** [reply]
Just want to acknowledge your positive reply. I'll follow things up when I have time. Thank you.

**Comment by kuiperzone** [comment]
Hi @kekekeks 

I was a little confused about the disadvantages you gave about an "in-process XAML loader", because I don't have [just] an in-process XAML loader (although it can do that as well). I understand you had many issues to overcome because I have faced them also, especially regarding Avalonia versioning.

However, I recognise there is a draw back to how I load an "out-of-process" preview. Essentially, the following must be inserted into the target project in order to resolve Avalonia versioning:

`<PackageReference Condition="'$(Configuration)' == 'Debug'" Include="AvantGarde.Services"/>`

And this assembly has its own main() function.

The vision for my app is that of a standalone previewer and not an extension to an IDE. I cannot and do not dispute the arguments you gave regarding the PreviewerProcess. However, if I had known that at the [very] start, I would have never have started my project because solving these problems were what grabbed my interest. In the process, I have learned a lot about XAML/Avalonia etc. because, while I am familiar with C#, I was not that familiar with WPF. So for me, this project is more about personal interest then anything else.

Now, I guess my question is:

**Is there any advantage or use-case to have an "in-process" XAML previewer which can load "code-behind" XAML without the assembly?** With this, it can locate and display resources and assets, but when faced with classes in other assemblies, such as UserControls, my "in-process" previewer stubs them out.

If the answer is no, there is no use case for this, then I have to consider stripping out much of my code.

I have 3 choices:

1. Continue with what I've got now

2 Leave in the "in-process" part of my code, which is less than trivial, so that it can load previews on its own without the assembly as described. But also add code to use your PreviewProcess to handle with-assembly loading.

3. Completely strip out almost of my own previewing code, and rely solely on your PreviewProcess.

Not sure what I'm going to do yet. But would be interested to know your reply.
